plugins/code-versions/src/utils/diff/line-diff.ts

```diff
@@ -0,0 +1,143 @@
+import { diffLines } from "diff"
+import { addContextLimitingAndDividers } from "./context-limiter"
+import { createAddLine, createChangeLine, createContextLine, createRemoveLine } from "./line-creators"
+import type { LineDiff } from "./types"
+
+/**
+ * Result of converting a diff block to line diffs.
+ *
+ * @param diffs - Array of line diff objects
+ * @param oldLine - Updated line number for the old file
+ * @param newLine - Updated line number for the new file
+ * @param skipNext - When true, indicates that the next diff block was already processed
+ *                   as part of a paired remove/add operation, so it should be skipped
+ *                   to avoid double-processing the same content.
+ */
+interface DiffBlockResult {
+    diffs: LineDiff[]
+    oldLine: number
+    newLine: number
+    skipNext: boolean
+}
+
+/**
+ * Generates a line-by-line diff between two strings with context limiting and dividers.
+ *
+ * @param oldStr - The original string to compare
+ * @param newStr - The revised string to compare against
+ * @returns An array of LineDiff objects representing the differences
+ */
+export function getLineDiff(oldStr: string, newStr: string): LineDiff[] {
+    const rawDiffs = diffLines(oldStr, newStr)
+    const result: LineDiff[] = []
+    let oldLine = 1
+    let newLine = 1
+
+    for (let i = 0; i < rawDiffs.length; i++) {
+        const currentDiff = rawDiffs[i]
+        if (!currentDiff) continue
+
+        const nextDiff = rawDiffs[i + 1]
+        const {
+            diffs,
+            oldLine: updatedOldLine,
+            newLine: updatedNewLine,
+            skipNext,
+        } = convertDiffBlockToLineDiffs(currentDiff, nextDiff, oldLine, newLine)
+
+        result.push(...diffs)
+        oldLine = updatedOldLine
+        newLine = updatedNewLine
+        // If skipNext is true, it means we processed both current and next diff blocks together
+        // as a paired remove/add operation, so we need to skip the next iteration to avoid
+        // processing the same block twice.
+        if (skipNext) i++
```

Comment:
Nit: I think we have a soft rule that only return/continue/break/throw can be inline, the rest in a block. Also, I'd encourage `i += 1`, slightly less error-prone 😛